### PR TITLE
fix(source): fix argument boundaries when parsing cmdline

### PR DESCRIPTION
**Pull request by gilbsgilbs**
## Proposed Commit Message

```
fix(source): fix argument boundaries when parsing cmdline

When the command-line includes an argument that contains `ds=`, the rest
of the argument is always interpreted as a datasource, regardless of
wether the argument is `ds` or just the suffix of another argument.
While this doesn't cause issues in a majority of cases, it starts
conflicting with other arguments that also include `ds=` but do not
refer to a cloud-init datasource. For instance, setting
`bonding.max_bonds=0` in the command-line (which is a real-world
use-case) conflicts with the current implementation.

To fix this issue, this commit ensures that `ds`, `ci.ds` and
`ci.datasource` arguments are either at the very beginning of the
command-line, or just before a whitespace, so that we don't match
another argument.
```

## Test Steps

Add `bonding.max_bonds=0` to the command-line, and then `ds=nocloud-net;s=https://some`. Result: the parsed datasource is `0` instead of `nocloud-net;s=https://some`.

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] (irrelevant as it is a bugfix) I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by holmanb** [comment]
Hey @gilbsgilbs, thanks for proposing this contribution. Can you please [sign the CLA](https://ubuntu.com/legal/contributors) so that we may accept this contribution?

**Comment by gilbsgilbs** [comment]
I'd like to, but it's asking me for a launchpad id that I can't manage to find it on my launchpad account.

File: cloudinit/sources/__init__.py [hunk 1]
```diff
@@ -1185,9 +1185,9 @@ def parse_cmdline() -> str:
     Passing by command line overrides runtime datasource detection
     """
     cmdline = util.get_cmdline()
-    ds_parse_0 = re.search(r"ds=([^\s;]+)", cmdline)
-    ds_parse_1 = re.search(r"ci\.ds=([^\s;]+)", cmdline)
-    ds_parse_2 = re.search(r"ci\.datasource=([^\s;]+)", cmdline)
+    ds_parse_0 = re.search(r"(?:^|\s)ds=([^\s;]+)", cmdline)
```
Comment (holmanb) [inline cloudinit/sources/__init__.py:1188]: The `?:` is meaningless in these regexes, did you intend something else?

**Review by holmanb — CHANGES_REQUESTED** [review]
@gilbsgilbs Great, thanks for signing! Can you please include in this PR your github username in [this file](https://github.com/canonical/cloud-init/blob/main/tools/.github-cla-signers)?

I also have one question about the regex.

File: cloudinit/sources/__init__.py (hunk 1 above)
Comment (gilbsgilbs) [reply cloudinit/sources/__init__.py:1188]: It makes the group non-capturing. The other group in these regexes is referred to as `.group(1)`, so if you capture the first group, you have to shift the group index. As the first group is only used to match what we want, I think making it non-capturing is "more correct" than shifting the group at the end of this function.

**Comment by gilbsgilbs** [comment]
> @gilbsgilbs Great, thanks for signing! Can you please include in this PR your github username in [this file](https://github.com/canonical/cloud-init/blob/main/tools/.github-cla-signers)?

Done.

**Review by holmanb — APPROVED** [review]
Thanks for the bug fix and welcome to cloud-init @gilbsgilbs!

**Pull request merged**